### DISCUSSION TITLE
CSP eval() blocked error in NextJS v9.5.3

### DISCUSSION THREAD

**Discussion by amit08255** [vercel/next.js#17396] [Help]
# Bug report

## Describe the bug

I am using NextJS v9.5.3. It was working perfectly before I updated to latest version. Now I am getting errors and in development mode the page is always blank. It works in production mode only with error from browser.

Following is the error I get -

```js
Content Security Policy: Couldn’t process unknown directive ‘script-src-attr’
Content Security Policy: The page’s settings blocked the loading of a resource at inline (“script-src”). injectGlobalHook.js:513:49
Content Security Policy: The page’s settings blocked the loading of a resource at inline (“script-src”). content-fontface.js:50:25
Content Security Policy: The page’s settings blocked the loading of a resource at inline (“script-src”). content-hooks-frames.js:62:42
Uncaught EvalError: call to eval() blocked by CSP
    NextJS 6
react-refresh.js:24:1
Uncaught EvalError: call to eval() blocked by CSP
    NextJS 8
_app.js:657:1
Uncaught EvalError: call to eval() blocked by CSP
    NextJS 5
login.js:658:1
Content Security Policy: The page’s settings blocked the loading of a resource at eval (“script-src”). react-refresh.js:24
Content Security Policy: The page’s settings blocked the loading of a resource at eval (“script-src”). _app.js:657
Content Security Policy: The page’s settings blocked the loading of a resource at eval (“script-src”).
```

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Install latest version of NextJS and build a simple page.
2. Run ```sh npm run dev``` in terminal.
3. Open browser and in console it will show the error and page will be all blank always.

## Expected behavior

Page should be loading and working perfectly without any error which is occuring now.

## Screenshots

![Screenshot from 2020-09-27 22-58-24](https://user-images.githubusercontent.com/28493237/94371577-2f091f00-0115-11eb-9766-414f4b5b0a95.png)

## Additional context

It was working perfectly in older version of NextJS (before I switched to v9.4+)

**Comment by amit08255** [comment]
This weird error has stopped my app completely. Still no support from this team. Worst issue ever had with NextJS

**Reply by bozdoganCihangir** [reply]
Did you find a solution? I am having the same error and last few hours i did so many changes in the codebase, i have no idea about where this error came up.

**Comment by bozdoganCihangir** [comment]
All works fine after removing the "helmet" npm package.

**Reply by imdoingotherthings** [reply]
This worked for me. 

It's an interesting error though, because I'm working on a project on 2 different OS's (MacOS and Windows 10 Home) and on Windows I have no problem running the NPM package `helmet` but when I go to work on MacOS, I get the `eval` error.

I just removed  `helmet` altogether and I'm looking to write a script to check if in production mode or dev mode for `helmet`.

**Reply by asitanc** [reply]
Thanks for pointing me to the right direction.
I have noticed one of my projects was fine, while other was causing CSP error. I checked the `helmet` version and noticed that with `3.23.0` it works fine, while with `5.1.0` it doesn't (did not work with `4.X.X` either for me).

**Comment by droid95** [comment]
It's because, as you said, you are running your app in development mode, so evals functions and some script src's is used in dev environment to help you in certain tasks. Build your app and run in production mode, you'll see all these eval errors not showing on console, because eval functions is not used in production mode.

**Comment by JustFly1984** [comment]
There is no point in running CSP in dev mode. Only if you are highly paranoid about your dev environment.

**Reply by bozdoganCihangir** [reply]
🤣
